ping/src/lib.rs

```diff
@@ -130,86 +148,119 @@ where
     ) -> Self::Future {
-        // # How does it work?
-        //
-        // All the actual processing is performed by the *ponger*.
-        // We use a channel in order to send ping requests from the pinger to the ponger.
-
-        let (tx, rx) = mpsc::channel(8);
-        // Ignore the errors if `tx` closed. `tx` is only ever closed if the ponger is closed,
-        // which means that the connection to the remote is closed. Therefore we make the `rx`
-        // never produce anything.
-        let rx = rx.then(|r| Ok(r.ok())).filter_map(|a| a);
-
-        let pinger = Pinger {
-            send: tx,
-            rng: EntropyRng::default(),
+        let out = match endpoint {
+            Endpoint::Dialer => upgrade_as_dialer(socket),
+            Endpoint::Listener => upgrade_as_listener(socket),
         };
 
-        // Hashmap that associates outgoing payloads to one-shot senders.
-        // TODO: can't figure out how to make it work without using an Arc/Mutex
-        let expected_pongs = Arc::new(Mutex::new(HashMap::with_capacity(4)));
-
-        let sink_stream = Framed::new(socket, Codec).map(|msg| Message::Received(msg.freeze()));
-        let (sink, stream) = sink_stream.split();
-
-        let future = loop_fn((sink, stream.select(rx)), move |(sink, stream)| {
-            let expected_pongs = expected_pongs.clone();
-
-            stream
-                .into_future()
-                .map_err(|(err, _)| err)
-                .and_then(move |(message, stream)| {
-                    let mut expected_pongs = expected_pongs.lock();
-
-                    if let Some(message) = message {
-                        match message {
-                            Message::Ping(payload, finished) => {
-                                // Ping requested by the user through the `Pinger`.
-                                debug!("Sending ping with payload {:?}", payload);
-
-                                expected_pongs.insert(payload.clone(), finished);
-                                Box::new(
-                                    sink.send(payload)
-                                        .map(|sink| Loop::Continue((sink, stream))),
-                                )
+        Ok((out, remote_addr)).into_future()
+    }
+}
+
+/// Upgrades a connection from the dialer side.
+fn upgrade_as_dialer(socket: impl AsyncRead + AsyncWrite + 'static) -> PingOutput {
+    // # How does it work?
+    //
+    // All the actual processing is performed by the *ponger*.
+    // We use a channel in order to send ping requests from the pinger to the ponger.
+
+    let (tx, rx) = mpsc::channel(8);
+    // Ignore the errors if `tx` closed.
+    let rx = rx.then(|r| Ok(r.ok())).filter_map(|a| a);
+
+    let pinger = Pinger {
+        send: tx,
+        rng: EntropyRng::default(),
+    };
+
+    // Hashmap that associates outgoing payloads to one-shot senders.
+    // TODO: can't figure out how to make it work without using an Arc/Mutex
```

Comment:
Are pongs not expected to be received in the same order as the pings were send? We aren't checking for that at the moment (they can respond in whichever order they want), but if that was implied by the protocol, we could simplify this codebase and use a FIFO-Buffer of expected pongs and have a `futures` per ping interface instead. However, not necessarily part of this PR as it is only touching the code base but concerned with other aspects of the protocol.